### PR TITLE
Lily/xlm roberta

### PR DESCRIPTION
**Pull request by lhz1029**

### CODE REVIEW

File: docker/Dockerfile.cpu [hunk 1]
```diff
@@ -4,7 +4,7 @@ LABEL Author="Ben Townsend, Madison May"
 # tensorboard
 EXPOSE 6006
 
-RUN apt-get update && mkdir /Finetune
+RUN apt-get update && apt-get install node.js && apt-get install npm && mkdir /Finetune
```
Comment (lhz1029) [inline docker/Dockerfile.cpu]: these are needed to import huggingface's tokenizer library. Given the extra weight, I'd probably suggest not merging this PR unless we find xlm-roberta useful enough to warrant it.

File: finetune/base_models/huggingface/models.py [hunk 2]
```diff
@@ -6,13 +6,100 @@
...
+        # The first "real" token "," has position 4 in the original fairseq vocab and position 3 in the spm vocab
+        self.fairseq_offset = 1
+
+        self.fairseq_tokens_to_ids["<mask>"] = len(self.sp_model) + self.fairseq_offset
+        self.fairseq_ids_to_tokens = {v: k for k, v in self.fairseq_tokens_to_ids.items()}
+
+
+
+HFXLMRoberta = finetune_model_from_huggingface(
+    pretrained_weights="jplu/tf-xlm-roberta-base",
```
Comment (lhz1029) [inline finetune/base_models/huggingface/models.py:24]: xlm-roberta-base does not have a tf version, so I'm using a model someone else ported

File: finetune/base_models/huggingface/models.py [hunk 3]
```diff
@@ -6,13 +6,100 @@
...
+from transformers.tokenization_utils import PreTrainedTokenizerFast
+from tokenizers import SentencePieceBPETokenizer
 
 from finetune.util.huggingface_interface import finetune_model_from_huggingface
 
 
+class XLMRobertaTokenizerFast(XLMRobertaTokenizer, PreTrainedTokenizerFast):
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+    # pretrained_vocab_files_map['merges_file'] = {"xlm-roberta-base": "/path"}
```
Comment (lhz1029) [inline finetune/base_models/huggingface/models.py]: I think I want these two lines uncommented, but I'm not completely positive if either configuration yields the exact same tokenizer as the original xlm-roberta (need to confirm with the tests).

The reason why I'm a bit unsure is that the SentencePieceBPETokenizer needs a vocab_file and a merges_file:
https://github.com/huggingface/tokenizers/blob/v0.1.1/bindings/python/tokenizers/implementations/sentencepiece_bpe.py#L21
(logic to use these files is in rust here: https://github.com/huggingface/tokenizers/blob/58e29fd30c4f0b163f145b885732d970a3c387bc/tokenizers/src/models/bpe/model.rs#L249)

The two are typically a *vocab.json and *merges.txt, but in jplu/tf-xlm-roberta-base's case, the vocab_file is a .model file which sentencepiece can take in directly:
https://huggingface.co/transformers/_modules/transformers/tokenization_xlm_roberta.html

File: finetune/base_models/huggingface/models.py [hunk 4]
```diff
@@ -6,13 +6,100 @@
...
+        self.fairseq_offset = 1
+
+        self.fairseq_tokens_to_ids["<mask>"] = len(self.sp_model) + self.fairseq_offset
+        self.fairseq_ids_to_tokens = {v: k for k, v in self.fairseq_tokens_to_ids.items()}
+
+
+
+HFXLMRoberta = finetune_model_from_huggingface(
+    pretrained_weights="jplu/tf-xlm-roberta-base",
+    archive_map={'jplu/tf-xlm-roberta-base': "https://s3.amazonaws.com/models.huggingface.co/bert/jplu/tf-xlm-roberta-large/tf_model.h5"},
```
Comment (benleetownsend) [inline finetune/base_models/huggingface/models.py]: Base -> Large?

Comment (lhz1029) [reply finetune/base_models/huggingface/models.py]: ah, thank you for catching that!

File: tests/test_huggingface.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+class TestHuggingFace(unittest.TestCase):
+    def setUp(self):
+        self.text = "The quick brown fox jumps over the lazy dog"
+
+    def check_embeddings_equal(self, finetune_base_model, hf_model_path):
+        finetune_model = Classifier(
+            base_model=finetune_base_model, train_embeddings=False, n_epochs=1)
+        finetune_model.fit([self.text], ['class_a'])
+        np.testing.assert_array_almost_equal(
+            finetune_model.featurize_sequence(self.text),
```
Comment (benleetownsend) [inline tests/test_huggingface.py]: `[self.text]`

File: tests/data/doc_with_empty_pages.json [hunk 6]
```diff
@@ -0,0 +1 @@
...tart": 7, "end": 12}, "page_offset": {"start": 6, "end": 11}, "block_offset": {"start": 6, "end": 11}, "position": {"top": 55, "bottom": 78, "left": 251, "right": 333, "bbBot": 80, "bbTop": 55, "bbLeft": 251, "bbRight": 333}, "page_num": 1, "style": {"bold": true, "underlined": false, "italic": false, "font_face": "", "font_size": 156, "background_color": "ffffff", "text_color": "000000"}}, {"text": "Care", "doc_offset": {"start": 13, "end": 17}, "page_offset": {"start": 12, "end": 16}, "block_offset": {"start": 12, "end": 16}, "position": {"top": 55, "bottom": 78, "left": 343, "right": 408, "bbBot": 80, "bbTop": 55, "bbLeft": 343, "bbRight": 408}, "page_num": 1, "style": {"bold": true, "underlined": false, "italic": false, "font_face": "", "font_size": 156, "background_color": "ffffff", "text_color": "000000"}}]}, {"pages": [{"text": "", "size": {"height": 2112, "width": 1408}, "dpi": {"dpix": 150, "dpiy": 150}, "doc_offset": {"start": 18, "end": 18}, "page_num": 2, "image": {"url": "indico-file:///storage/uploads/12404/2030bca6-5347-4ae1-ba58-e8cbc20175fc.png"}, "thumbnail": {"url": "indico-file:///storage/uploads/12404/dfac991b-bdf0-4f30-8097-7db41ffe1838.jpg"}}], "tokens": []}]
```
Comment (lhz1029) [inline tests/data/doc_with_empty_pages.json:1]: this should have been added in a previous PR I created

File: finetune/util/huggingface_interface.py [hunk 7]
```diff
@@ -127,28 +130,53 @@ def __init__(self):
...
+                    batch_token_idxs.append(encoded.ids)
+                    token_ends = []
+                    token_starts = []
+                    for start, end in encoded.offsets:
+                        token_starts.append(start)
+                        token_ends.append(end)
+
+                    batch_char_ends.append(token_ends)
+                    batch_char_starts.append(token_starts)
+                else:
```
Comment (madisonmay) [inline finetune/util/huggingface_interface.py:155]: Awesome, thank you for taking the time to support the non-fast tokenizers w/o char offsets here

**Review by madisonmay — APPROVED** [review]

File: finetune/util/huggingface_interface.py (hunk 7 above)
Comment (lhz1029) [reply finetune/util/huggingface_interface.py:155]: Of course! In retrospect I should have done this first haha....

Comment (benleetownsend) [inline finetune/util/huggingface_interface.py:155]: Is there some way to detect the SentencePiece based tokenizers?

File: tests/test_huggingface.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+from finetune import SequenceLabeler
+from finetune.base_models.huggingface.models import HFBert, HFElectraGen, HFElectraDiscrim, HFXLMRoberta
+
+
+class TestHuggingFace(unittest.TestCase):
+    def setUp(self):
+        self.text = """
+        The quick brown fox jumps over the lazy dog. Il ne faut pas se fier aux apparences.
+        Año nuevo, vida nueva. Das ist mir Wurst.
```
Comment (benleetownsend) [inline tests/test_huggingface.py]: Might be worth including this tokenizer in the tokenizer test? Especially if we plan on using it on languages with more complicated alphabets. test.https://github.com/IndicoDataSolutions/finetune/blob/c04a6d01ba8568c26c2d201c2dba5385027b0492/tests/test_utils.py#L35

File: finetune/util/tokenization.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
+import unicodedata
+
+
+WEIRD_SPM_CHAR = "▁"
+WEIRD_TEXT = """This is a basic test of tokenizers but with some fun things at the end.
```
Comment (madisonmay) [inline finetune/util/tokenization.py]: Can we move this constant to a test file?

Comment (lhz1029) [reply finetune/util/tokenization.py]: yep!

**Pull request merged**